### PR TITLE
replace with getField

### LINKED ISSUE #7835: [BUG] FlatObjectFieldMapper calls getFields, not getField
**Describe the bug**

In FlatObjectFieldMapper#parseValueAddFields, getFields is called, but getField needs to be used.
(In our performance tests, this issue reduced indexing speed.)

**To Reproduce**

Please see [FlatObjectFieldMapper.java#L631](https://github.com/opensearch-project/OpenSearch/blob/2.7.0/server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java#L631).

**Expected behavior**

Please replace _getFields_ with _getField_.

**Plugins**

None

**Screenshots**

None

**Host/Environment (please complete the following information):**

 - OS: OS-Independent
 - Version 2.7

**Additional context**

This issue was reported based on this [comment](https://forum.opensearch.org/t/flatobjectfieldmapper-calls-getfields-not-getfield/14451/1) from the Forum.

### PR DESCRIPTION
**Pull request by marevol**
### Description
In FlatObjectFieldMapper#parseValueAddFields, getFields is called, but getField needs to be used.

### Related Issues
Resolves #7835 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:634]: Super minor, I think you could simplify it a bit with:

```suggestion
                if (!Objects.equals(context.doc().getField(fieldType().name()), field)) {
```

**Comment by reta** [comment]
Thanks for the change, @marevol , it would be great to have a test case that showcased this issue (and the fix), could you craft one please? Thank you.

**Comment by marevol** [comment]
Our test case is:

1) Preparing 20,000,000 documents containing flat_object data
2) Inserting the documents and measuring the indexing time
   (we compared it with the flattened type in Elasticsearch. The indexing time for the flattened type was 3 hours, but for flat_object, it was 12 hours...)
3) Checking the result from the Nodes hot threads API
    'getFields' appears to be time-consuming in the results.

**Comment by reta** [comment]
> Our test case is:

@marevol  That's certainly a good one, do you see the way the issue could have been caught in unit / integration tests since it obviously does use wrong API? thank you.

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:634]: It needs import please:

```
import java.util.Objects;
```

**Comment by reta** [comment]
Please add the entry to CHANGELOG file, under [Unreleased 2.x] / Fixed section

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:635]: Fixed.
(I clicked the Suggestion Merge button on GitHub, but it did not work..)

Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:635]: It seems like it did? https://github.com/opensearch-project/OpenSearch/pull/7855/files#diff-8a5849247d51dba2b09b6505e299c618dbd3ca9b38d3ee1a30b07941c7bd7a9bR51

Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:635]: Oh, it was not the suggestion since the import section didn't change, sorry about the confusion

Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:635]: Thank you for checking it.
The "Suggestion Merge button" is GitHub feature. I clicked it without adding import...

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (noCharger) [inline server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:631]: Looks like this is a bug from the source code since `equals()` will always return `false`.

```
        public IndexableField[] getFields(String name) {
            List<IndexableField> f = new ArrayList<>();
            for (IndexableField field : fields) {
                if (field.name().equals(name)) {
                    f.add(field);
                }
            }
            return f.toArray(new IndexableField[0]);
        }

        public IndexableField getField(String name) {
            for (IndexableField field : fields) {
                if (field.name().equals(name)) {
                    return field;
                }
            }
            return null;
        }
```

**Review by noCharger — APPROVED** [review]

**Comment by noCharger** [comment]
> We found this issue in performance tests. The behavior of flat_object type looked okay. So it might be difficult to find it in JUnit tests.

Although the fix looks good to me, I'm wondering whether we may create a test case that yields several fields using `getFields()` but just one from `getField`. Thanks.

**Comment by reta** [comment]
> Although the fix looks good to me, I'm wondering whether we may create a test case that yields several fields using `getFields()` but just one from `getField`. Thanks.

I was trying to craft a test for it and it turned out to be difficult, primarily because the FlatObjectFieldType always returns `true` for doc values and due to the bug in initial code, the additional `SortedSetDocValuesField` was always added as well  ....

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: Yes, we can! And let's found out why test fails :)

Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: > If you move it to the previous place, JUnit tests will be failed. Can we move it?

@marevol because now with the if statement of getField will become true, it will add SortedSetDocValuesField into the doc(), and the indexableField length will be get changed in here in https://github.com/opensearch-project/OpenSearch/blob/28d5948d4fb5acd55fc686c05058c85b12b95c18/server/src/test/java/org/opensearch/index/mapper/FlatObjectFieldMapperTests.java#LL110C37-L110C37 You can try to print out the length and also the indexable fields and see if that's expected.

Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: Although I might not understand it correctly, if moving the code to the previous location like [this](https://github.com/marevol/OpenSearch/commit/415b4bbcea20afd994cf66a6d1b5b7fe77cc46bc), the length will be 1.
However, [the moved code](https://github.com/marevol/OpenSearch/blob/415b4bbcea20afd994cf66a6d1b5b7fe77cc46bc/server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java#LL634C23-L634C23) seems not to be called...

Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: Looking into the changed  code here: `!Objects.equals(context.doc().getField(fieldType().name()), field)` , it would return false, because the field should be the same. 

Then the SortedSetDocValuesField would not be added.  The indexable length will be 1 instead of 2.

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (mingshl) [inline server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:632]: In the parent field, flat_object field, when fieldName.equals(fieldType().name()), we don't need to store the SortedSetDocValueField, because the parent field stores only the key of the field name in the JSON object, it's used in checking `exist query`. 
But in other queries, the mapping will direct the query to its subfields, either VALUE field or VALUE_AND_PATH field. 

The parent field type is not intended to used for sort, aggregate, and perform other operations on those values, will be directed to subfield instead, so  we can remove L632 to L634.

The expected length of indexable fields in a field {"field: { "foo": "bar"}}"

the parent field should contain one IndexableField, and the subfields should contains two IndexableField, in which one of them is SortedSetDocValueField: 
```
        IndexableField[] fields = doc.rootDoc().getFields("field");
        assertEquals(1, fields.length);
        assertEquals(new BytesRef("field.foo"), fields[0].binaryValue());

        IndexableFieldType fieldType = fields[0].fieldType();
        assertFalse(fieldType.tokenized());
        assertFalse(fieldType.stored());
        assertThat(fieldType.indexOptions(), equalTo(IndexOptions.DOCS));
        assertEquals(DocValuesType.NONE, fieldType.docValuesType());

        // Test internal substring fields as well
        IndexableField[] fieldValues = doc.rootDoc().getFields("field" + VALUE_SUFFIX);
        assertEquals(2, fieldValues.length);
        assertTrue(fieldValues[0] instanceof KeywordFieldMapper.KeywordField);
        assertEquals(new BytesRef("bar"), fieldValues[0].binaryValue());

        IndexableField[] fieldValueAndPaths = doc.rootDoc().getFields("field" + VALUE_AND_PATH_SUFFIX);
        assertEquals(2, fieldValues.length);
        assertTrue(fieldValueAndPaths[0] instanceof KeywordFieldMapper.KeywordField);
        assertEquals(new BytesRef("field.foo=bar"), fieldValueAndPaths[0].binaryValue());
```

Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:632]: @reta @dblock @andrross @owaiskazi19 asking for opinions

Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:632]: @mingshl The `FlatObjectFieldType` as of today (please correct me if I am wrong) is always created with `hasDocValues` set `true`, that would mean that `existsQuery` would use `DocValuesFieldExistsQuery` (which seems to be deprecated in flavor of `FieldExistsQuery`). If the parent field type is not intended is not intended to store doc values, we could instantiate it with `hasDocValues` set `false` instead.

Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:632]: Agree that hasDocvalue is a bigger discussion. We should separate from this bug fix. I will raise a new issue related to the hasDocValue().

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: [marevol](https://github.com/marevol) does changing the indexable length to 1 help pass the test? can we get this bug fix before 2.9 code freeze on July 11?

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:631]: Agree, thanks @mingshl

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: Yes. Changing to 1 passes the test case. 
If [this fix](https://github.com/marevol/OpenSearch/commit/415b4bbcea20afd994cf66a6d1b5b7fe77cc46bc) looks okay, I'll update this PR by it.

Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java:619]: LGTM. Please update this PR. 

I will address the docValue in a different issue to open up a new discussion.

Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java]: Thank you for checking it. Updated.

**Comment by noCharger** [comment]
2> REPRODUCE WITH: ./gradlew ':qa:smoke-test-multinode:integTest' --tests "org.opensearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT.test {yaml=search/110_field_collapsing/field collapsing and search_after}" -Dtests.seed=4731AFD2E93B5B47 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Australia/Darwin -Druntime.java=20

https://build.ci.opensearch.org/job/gradle-check/19136/consoleFull

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java]: Other test cases failed... 
It might be better to revert dea6dbf9d537e1df80984536dd590372d3d252fb, merge this PR, and then fix the hasDocvalue problem as another problem.

**Comment by noCharger** [comment]
```
org.opensearch.painless.LangPainlessClientYamlTestSuiteIT > test {yaml=painless/30_search/Flat-object fields from within the scripting} FAILED
    java.lang.AssertionError: Failure at [painless/30_search:559]: field [hits.hits] doesn't have length [1]
    Expected: <1>
         but: was <0>
        at __randomizedtesting.SeedInfo.seed([DACE9185F5EE8DF:85F8D6C2F1A28527]:0)
        at org.opensearch.test.rest.yaml.OpenSearchClientYamlSuiteTestCase.executeSection(OpenSearchClientYamlSuiteTestCase.java:460)
        at org.opensearch.test.rest.yaml.OpenSearchClientYamlSuiteTestCase.test(OpenSearchClientYamlSuiteTestCase.java:433)
        at java.****/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.****/java.lang.reflect.Method.invoke(Method.java:578)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)

```

**Comment by mingshl** [comment]
@marevol  we can remove this painless from L488  [test]( https://github.com/opensearch-project/OpenSearch/blame/bcaf49455fc6dacd1e19e32b8457a8bc33991c7a/modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml#L488). It should help you with the gradle  ceck. 

I remembered @lukas-vlcek  was nice to add in this painless test. But since we agree on FlatObject currently do not support painless script, and we have a new issue assigned to lukas for future development, so it's fine to remove this test [here](https://github.com/opensearch-project/OpenSearch/issues/7138).

File: server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java
Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java]: I saw the hasDocValue tests failed, so  I tried on my fork, by simply removing the getField line [here](https://github.com/mingshl/OpenSearch-Mingshl/blob/2f9728e8398e923cbbb93e36dd52c22e02612a8d/server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java#L631), it adds docValues as it was, and the length of the indexable field is still 2 and the rest of the tests are working fine. 

before pushing the PR, try running `./gradlew precommit` and `./gradlew test`, in this way you can make sure all the tests passed. 

If we are proposing to set the FlatObject hasDocValue to false, we will raise a new PR. Let's make the bug fix first then thinking about changing the docvalue.

Comment (mingshl) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java]: @marevol do you have luck in trying the above approach?

Comment (marevol) [reply server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java]: Your change was merged into this branch, and Gradle checks have passed.

**Comment by mingshl** [comment]
@reta can we get reviews again? Trying to see if we can make this in for. 2. 9?

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml
Comment (reta) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml:488]: @mingshl apologies if I missed that, but why these test are removed?

Comment (mingshl) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml:488]: The flat object currently doesn't  support painless query yet. And it will be a future implementation.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@mingshl sign off from you before merge? thank you

**Review by mingshl — APPROVED** [review]

**Comment by reta** [comment]
@mingshl mind please backporting manually ? thank you

**Pull request merged**
